src/ast/visitor.h

```diff
@@ -0,0 +1,369 @@
+#pragma once
+
+#include <string>
+
+#include "ast/ast.h"
+
+namespace bpftrace::ast {
+
+// Visitor for fully-static visitation.
+//
+// This uses CRTP to make all calls static, while still allowing the entrypoint
+// for a single visitor to be dispatched dynamically. The implementation may
+// optionally provide individual `visit` methods (matching either pointers or
+// references, the latter preferred), or `replace` methods (matching just the
```

Comment:
Am I getting it right in that you intend to follow up this PR with one that moves `dereference_if_needed()` out into its own pass? And that would allow you to make `visit` method take const ref? Or do we still want `visit` to be able to mutate the AST?

Reply:
Visit is allowed to mutate the AST, it is just not allowed to replace a node in the graph. I didn't have plans otherwise, but it's actually an interesting idea to have *nothing* be allowed to mutate via visit (replace could also pass a const ref). The mutation would be passed up the graph automatically. But maybe this is getting a bit hairy and complicated, and I'd rather keep everything as simple as possible. WDYT?

Reply:
> Visit is allowed to mutate the AST, it is just not allowed to replace a node in the graph.

Not allowed to replace any node in the graph, or just not allowed to replace the _current_ node in the graph? Seems like it should be possible to replace children nodes, right?

Reply:
If I'm reading this right, `visit` is allowed to mutate and `replace` is allowed to completely replace. I think I would prefer if `visit` wasn't allowed to mutate and we only did explicit replacement instead of mutation otherwise I'm not sure why we should have both methods.

Reply:
Makes sense. Not a requirement for this PR, just thinking about the future state of things.